File: Homework/Hoang_Nguyen/lesson3/239_slidingWindowMaximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
  // stores index of elements in nums array
  const indexQueue = [];
  const result = [];
  for (let i = 0; i < nums.length; i++) {
    // remove all elements in queue that are smaller than current element
    while (
      indexQueue.length &&
      nums[indexQueue[indexQueue.length - 1]] < nums[i]
    ) {
      indexQueue.pop();
    }
    indexQueue.push(i);
    // reached the first sliding window
    if (i >= k - 1) {
      // dequeue invalid index in sliding window
      while (indexQueue[0] < i - k + 1) {
        indexQueue.shift();
      }
      result.push(nums[indexQueue[0]]);
    }
  }
  return result;
};

File: Homework/Ngan_Luu/239-sliding-window-maximum.py
https://leetcode.com/problems/sliding-window-maximum/
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        output = []
        queue=[]
        for i in range(len(nums)):
            while queue and nums[queue[-1]] <= nums[i]:
                queue.pop()
            queue.append(i)

            while queue and i - queue[0] >= k:
                queue.pop(0)

            if i >= k - 1:
                output.append(nums[queue[0]])

        return output